design-documents/web-api-controllers.md

```diff
@@ -0,0 +1,374 @@
+can have _property_ elements (which can have child _properties_ as well) to specify entity IDs and groups for those
+properties. The ability to specify these entity IDs/groups for properties explicitly instead of relying on auto-generated
+ones (like having _customer.addresses_ for addresses) is needed to reuse rules for different endpoints.
+ 
+When validation fails web API framework will throw ValidationException. This exception has to be introduced since existing
+Magento\Framework\Validator\Exception cannot describe validation errors properly. The exception will be rendered as following:
+ 
+```json
+{
+    "errorMessage": "Validation failed",
+    "validation_errors": {
+        "arguments": {
+            "customer": {
+                "general": {
+                    "customer_pwd_as_login": {"message": "Cannot have password being equal to login"}
+                },
+                "properties": {
+                    "email": {
+                        "general": {
+                          "string_length_min": {"message": "Length must be greater then 10", "min": 10},
+                          "customer_unique_email": {"message": "E-mail is already in use"}
+                        },
+                        "properties": {}
+                    },
+                    "addresses": {
+                        "general": {
+                            "collection_max": {"message": "Cannot contain more then 10 items", "max": 10}
+                        },
+                        "properties": {
+                            "0": {
+                                "general": {},
+                                "properties": {
+                                    "postcode": {
+                                        "general": {
+                                            "string_length_min": {"message": "Length must be greater then 5", "min": 5}
+                                        },
+                                        "properties": {}
+                                    }
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+}
+```
+ 
+Controllers and GraphQL resolvers would be able to call validators and pass them entity name/group name explicitly
+```php
+Magento\Framework\Validator\Factory::createValidator($entityName, $groupName, $builderConfig)
+```
+right before passing arguments to service contracts. This way validation rules could be reused and extended by 3rd-party
+developers via configs.
+ 
+This rules lists are extensible by 3rd party developers allowing them to easily add validation for new properties
+they introduce or change defaults via config files.
+ 
+ 
+#### Introduce reusable user input processors
+Sometimes before passing arguments to a service contracts we would have to perform user input processing that cannot
```

Comment:
This section should also be moved to dedicated Validation proposal. That proposal should describe validation approach holistically. This topic was discussed multiple times in the past but I don't think we have solid technical vision for validation approach.

Reply:
These user input processors are the topic of this design doc. They are not meant to be validators but rather classes containing user-input processing/presentation-layer logic to be reused among other presentation-layer classes like controllers/graphQL resolvers